WordPress/src/main/java/org/wordpress/android/ui/accounts/signup/UsernameChangerFullScreenDialogFragment.java

```diff
@@ -0,0 +1,322 @@
+import android.support.v7.widget.RecyclerView;
+import android.support.v7.widget.SimpleItemAnimator;
+import android.text.Editable;
+import android.text.Html;
+import android.text.Spanned;
+import android.text.SpannedString;
+import android.text.TextUtils;
+import android.text.TextWatcher;
+import android.view.ContextThemeWrapper;
+import android.view.LayoutInflater;
+import android.view.View;
+import android.view.ViewGroup;
+import android.widget.ProgressBar;
+import android.widget.TextView;
+
+import org.greenrobot.eventbus.Subscribe;
+import org.greenrobot.eventbus.ThreadMode;
+import org.wordpress.android.R;
+import org.wordpress.android.WordPress;
+import org.wordpress.android.analytics.AnalyticsTracker;
+import org.wordpress.android.fluxc.Dispatcher;
+import org.wordpress.android.fluxc.generated.AccountActionBuilder;
+import org.wordpress.android.fluxc.store.AccountStore.FetchUsernameSuggestionsPayload;
+import org.wordpress.android.fluxc.store.AccountStore.OnUsernameSuggestionsFetched;
+import org.wordpress.android.ui.FullScreenDialogFragment.FullScreenDialogContent;
+import org.wordpress.android.ui.FullScreenDialogFragment.FullScreenDialogController;
+import org.wordpress.android.ui.accounts.signup.UsernameChangerRecyclerViewAdapter.OnUsernameSelectedListener;
+import org.wordpress.android.util.ActivityUtils;
+import org.wordpress.android.util.AppLog;
+import org.wordpress.android.util.AppLog.T;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.inject.Inject;
+
+public class UsernameChangerFullScreenDialogFragment extends Fragment implements
+        FullScreenDialogContent, OnUsernameSelectedListener {
+    private ProgressBar mProgressBar;
+
+    protected FullScreenDialogController mDialogController;
+    protected Handler mGetSuggestionsHandler;
+    protected RecyclerView mUsernameSuggestions;
+    protected Runnable mGetSuggestionsRunnable;
+    protected String mDisplayName;
+    protected String mUsername;
+    protected String mUsernameSelected;
+    protected String mUsernameSuggestionInput;
+    protected TextInputEditText mUsernameView;
+    protected TextView mHeaderView;
+    protected UsernameChangerRecyclerViewAdapter mUsernamesAdapter;
+    protected boolean mShouldWatchText;  // Flag handling text watcher to avoid network call on device rotation.
+    protected int mUsernameSelectedIndex;
+
+    public static final String EXTRA_DISPLAY = "EXTRA_DISPLAY";
+    public static final String EXTRA_USERNAME = "EXTRA_USERNAME";
+    public static final String KEY_SHOULD_WATCH_TEXT = "KEY_SHOULD_WATCH_TEXT";
+    public static final String KEY_USERNAME_SELECTED = "KEY_USERNAME_SELECTED";
+    public static final String KEY_USERNAME_SELECTED_INDEX = "KEY_USERNAME_SELECTED_INDEX";
+    public static final String KEY_USERNAME_SUGGESTIONS = "KEY_USERNAME_SUGGESTIONS";
+    public static final String RESULT_USERNAME = "RESULT_USERNAME";
+    public static final int GET_SUGGESTIONS_INTERVAL_MS = 2000;
```

Comment:
Highly subjective: 2secs feel a bit much for my taste. Wdyt about making it 1s or around 1s?

Reply:
This is subjective and fairly arbitrary.  I chose a value that would try to ensure the user has stopped typing in order to avoid premature and frequent network calls.  It would be better if we had data to support that decision though.  @folletto, do you have any idea what a good time would be for this interaction?  This value is the amount of time in milliseconds to wait after the user has stopped typing to fetch the username suggestions from the API.

Reply:
The delay should be sub-1s. I think we use 250-300ms in search in Calypso, with debounce and sequencing the calls to avoid an old call to reset a newer call.

Reply:
I'm debounching the calls based on an (arbitrarily chosen) 400ms delay for the relevant constant in [site-creation domain selection screen](https://github.com/wordpress-mobile/WordPress-Android/pull/7155/files#diff-12fe24c901e61f13416d3cbfab63f6e4R29). I can adjust it to 250-300 if we decide to.

Reply:
> I think we use 250-300ms in search in Calypso

I would argue that we should not follow the desktop pattern for mobile in this case and avoid making multiple network calls in quick succession for a few reasons.
1. Metered data
2. Network speed
3. Network reliability
4. Typing speed
5. Typing reliabiliy

Waiting 2000ms might be too slow, but 300ms feels way too fast.  @hypest's suggestion of 1000ms or more might be a good compromise.

Reply:
I disagree here. Calypso has all the same considerations for mobile web. That timing is also a timing based on human expectation of a response, and bounce+sort take care of the networking aspect regardless of the connection.

Reply:
I updated the value in ec00d88581419424c927f4c3d94677d74d774a41.